base/reflection.jl

```diff
@@ -1963,7 +1963,7 @@ Behaves like `Core.Typeof`, except that a type with a free typevar does not get
 turned into `Type` and is thus valid in a signature without further UnionAll
 rewrapping.
 """
-TypeofValid(@nospecialize(x)) = isa(x, Type) && !has_free_typevars(x) ? Type{x} : typeof(x)
+TypeofValid(@nospecialize(x)) = x isa Type && x isa Type{x} ? Type{x} : typeof(x)
```

Comment:
Shouldn't `has_free_typevars` be faster than `isa T{x}` (assuming we mark the former ccall as pure)? I realize `jl_isa` does implement that check, but I am less sure if inference and subtyping would be willing to form `Const(Type{x})` when `x` is not valid or if it will primarily attempt to only use `Const(x)` and `typeof(x)` but would avoid forming `Type{x}` to be able to check this during inference.

Reply:
I tried to mark `has_free_typevars` as foldable and `@pure`, but `TypeofValid(Array)` is still unstable.
No sure what makes this difference. But I think the extra overhead looks negligible, and the stability of invalid `x` doesn't matter in real cases?